Modules/Sources/Yosemite/PointOfSale/OrderList/PointOfSaleOrdersProductImageResolver.swift

```
@@ -0,0 +1,55 @@
+
+public protocol PointOfSaleOrdersProductImageResolverProtocol {
+    func resolveImages(for order: POSOrder) async -> POSOrder
+    func resolveImagesFromCache(for order: POSOrder) async -> POSOrder
+}
+
+/// Downloads product image URLs for POS orders, utilizing caching and batch loading for efficiency.
+/// Preferably, in the future API calls have to be replaced by the local product database access.
+///
+public actor PointOfSaleOrdersProductImageResolver: PointOfSaleOrdersProductImageResolverProtocol {
+    private var imageCache: [Int64: String] = [:]
+    private let productsRemote: ProductsRemoteProtocol
+    private let siteID: Int64
+
+    public init(
+        siteID: Int64,
+        credentials: Credentials?
+    ) {
+        self.siteID = siteID
+        self.productsRemote = ProductsRemote(network: AlamofireNetwork(credentials: credentials))
+    }
+
+    /// Resolves images for all order items by fetching missing images and then using the cache.
+    public func resolveImages(for order: POSOrder) async -> POSOrder {
+        let uncachedProductIDs = Set(order.lineItems.map(\.productID)).subtracting(imageCache.keys)
+
+        if !uncachedProductIDs.isEmpty {
+            await fetchAndCacheProductImages(for: Array(uncachedProductIDs))
+        }
+
+        return await resolveImagesFromCache(for: order)
+    }
+
+    /// Resolves images for all order items using only the data currently in the cache.
+    public func resolveImagesFromCache(for order: POSOrder) async -> POSOrder {
+        let updatedItems = order.lineItems.map { item in
+            item.copy(imageURL: .some(imageCache[item.productID]))
+        }
+        return order.copy(lineItems: .some(updatedItems))
+    }
+
+    /// Fetches product data and updates the cache.
+    private func fetchAndCacheProductImages(for productIDs: [Int64]) async {
+        guard !productIDs.isEmpty else { return }
+
+        let products = try? await productsRemote.loadProducts(for: siteID, by: productIDs)
```

Should we wrap this in a do/catch block and log/bubble-up if there's an error? Otherwise might fail silently from throwing on the remote.

---

I wasn't sure if there was value in even logging some network/server failure 🤔 Of course, we could add it to help out with troubleshooting some issues where users don't see photos for their order products.